src/sinks/statsd.rs

```diff
@@ -0,0 +1,358 @@
+use crate::{
+    buffers::Acker,
+    event::{metric::Direction, Event, Metric},
+    sinks::util::{BatchServiceSink, Buffer, SinkExt},
+    topology::config::{DataType, SinkConfig},
+};
+use futures::{future, sink::Sink, Future, Poll};
+use serde::{Deserialize, Serialize};
+use snafu::{ResultExt, Snafu};
+use std::collections::HashMap;
+use std::net::{IpAddr, Ipv4Addr, SocketAddr, UdpSocket};
+use std::time::Duration;
+use tower::{Service, ServiceBuilder};
+
+#[derive(Debug, Snafu)]
+enum BuildError {
+    #[snafu(display("failed to bind to udp listener socket, error = {:?}", source))]
+    SocketBindError { source: std::io::Error },
+}
+
+pub struct StatsdSvc {
+    client: Client,
+}
+
+pub struct Client {
+    socket: UdpSocket,
+    address: SocketAddr,
+}
+
+impl Client {
+    pub fn new(address: SocketAddr) -> crate::Result<Self> {
+        let from = SocketAddr::new(IpAddr::V4(Ipv4Addr::new(127, 0, 0, 1)), 0);
+        let socket = UdpSocket::bind(&from).context(SocketBindError)?;
+        Ok(Client { socket, address })
+    }
+
+    pub fn send(&self, buf: &[u8]) -> usize {
+        self.socket
+            .send_to(buf, &self.address)
+            .map_err(|e| error!("error sending datagram: {:?}", e))
+            .unwrap_or_default()
+    }
+}
+
+#[derive(Deserialize, Serialize, Debug, Clone)]
+#[serde(deny_unknown_fields)]
+pub struct StatsdSinkConfig {
+    pub namespace: String,
+    #[serde(default = "default_address")]
+    pub address: SocketAddr,
+    pub batch_size: Option<usize>,
+    pub batch_timeout: Option<u64>,
+}
+
+pub fn default_address() -> SocketAddr {
+    SocketAddr::new(IpAddr::V4(Ipv4Addr::new(127, 0, 0, 1)), 8125)
+}
+
+#[typetag::serde(name = "statsd")]
+impl SinkConfig for StatsdSinkConfig {
+    fn build(&self, acker: Acker) -> crate::Result<(super::RouterSink, super::Healthcheck)> {
+        let sink = StatsdSvc::new(self.clone(), acker)?;
+        let healthcheck = StatsdSvc::healthcheck(self.clone())?;
+        Ok((sink, healthcheck))
+    }
+
+    fn input_type(&self) -> DataType {
+        DataType::Metric
+    }
+}
+
+impl StatsdSvc {
+    pub fn new(config: StatsdSinkConfig, acker: Acker) -> crate::Result<super::RouterSink> {
+        // 1432 bytes is a safe default to fit into MTU
+        // https://github.com/statsd/statsd/blob/master/docs/metric_types.md#multi-metric-packets
+        // also one might keep an eye on server side limitations, like
+        // https://github.com/DataDog/dd-agent/issues/2638
+        let batch_size = config.batch_size.unwrap_or(1432);
```

Comment:
Actually, thinking about it more, this could be a small issue. The way that our batch sink and buffer work, this is going to be the _minimum_ size of the buffer before flushing. So the actual sizes will almost always be larger. Should we bump this down a bit to try to keep the packet sizes under 1432?

Reply:
We'd still have the same issue in that case because [we don't check if the new item being added to the batch will cause us to exceed our max size](https://github.com/timberio/vector/blob/d35ddfff2edbc4f776a75cc420f834a6f4d2aec4/src/sinks/util/batch.rs#L129-L149).

We could do some work to the `BatchSink` to enable that kind of check, but up to this point [we've explicitly been ok with exceeding the max by one item in cases like this](https://github.com/timberio/vector/blob/d35ddfff2edbc4f776a75cc420f834a6f4d2aec4/src/sinks/util/batch.rs#L235).

Reply:
I see. Another option could be splitting the data into two packets before sending, but I'm perfectly fine with updating the default batch size.

Reply:
Yeah, I think reducing the default batch size is an easy fix for now.